### PR TITLE
nixos/technitium-dns-server: init module

### PR DESCRIPTION
**Pull request by FabianRig**
## Description of changes

This PR initializes a module for the Technitium DNS Server.

- Package has been added with https://github.com/NixOS/nixpkgs/pull/293326
- Info on Technitium DNS Server: https://technitium.com/dns/
- GitHub repo: https://github.com/TechnitiumSoftware/DnsServer

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by GiggleSquid — APPROVED** [review]
This is great to see! I've been meaning to work on this myself for months and just haven't found the time.

The module looks great, and does everything I'd expect it to.
Thanks for this  (and the pkg) :)

**Comment by FabianRig** [comment]
Thank you very much for your review and the kind words!

I was surprised, that there was no package for Technitium -- best DNS server out there. ;-)

File: nixos/modules/services/networking/technitium-dns-server.nix [hunk 1]
```diff
@@ -0,0 +1,130 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+with lib;
```
Comment (Aleksanaa) [inline nixos/modules/services/networking/technitium-dns-server.nix:8]: Don't use top-level `with lib`, use `inherit` instead.

File: nixos/modules/services/networking/technitium-dns-server.nix [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+{
+  options.services.technitium-dns-server = {
+    enable = mkEnableOption "Technitium DNS Server";
+
+    package = mkPackageOption pkgs "technitium-dns-server" { };
+
+    openFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = mdDoc ''
```
Comment (Aleksanaa) [inline nixos/modules/services/networking/technitium-dns-server.nix:23]: mdDoc is no longer needed as it has become the default format. You can remove all of them safely.

File: nixos/modules/services/networking/technitium-dns-server.nix (hunk 1 above)
Comment (FabianRig) [reply nixos/modules/services/networking/technitium-dns-server.nix]: Thank you for your review. I hadn't seen that before. Fixed it.

File: nixos/modules/services/networking/technitium-dns-server.nix (hunk 2 above)
Comment (FabianRig) [reply nixos/modules/services/networking/technitium-dns-server.nix]: All mdDoc removed. Thanks again!

**Review by adamcstephens — COMMENTED** [review]
It would be good to write a nixos test for the module and add it to passthru.tests for the package.

**Comment by FabianRig** [comment]
> It would be good to write a nixos test for the module and add it to passthru.tests for the package.

Tbh, I currently don't know how to do that. I'd prefer to add tests in a future iteration once I have more time to look into tests.

File: nixos/modules/services/networking/technitium-dns-server.nix [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  config = mkIf cfg.enable {
+    systemd.services.technitium-dns-server = {
+      description = "Technitium DNS Server";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/technitium-dns-server ${stateDir}";
+
+        User = cfg.user;
```
Comment (drupol) [inline nixos/modules/services/networking/technitium-dns-server.nix:82]: How about using a dynamic user and removing the system user creation?

**Review by drupol — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/technitium-dns-server.nix (hunk 3 above)
Comment (FabianRig) [reply nixos/modules/services/networking/technitium-dns-server.nix]: That simplifies the module a lot. My first impression was that a dedicated user would be better. I have changed that to a dynamic user.

**Comment by FabianRig** [comment]
Unfortunately, one of the check is failing now. I assume that it's *not* an error on my end?

~~~
Binary cache nixpkgs-ci doesn't exist or it's private and you need a token: Start by visiting https://app.cachix.org/ and create a personal/cache auth token.
  
  To configure the token:
  
  a) Via environment variable: 
  
  $ export CACHIX_AUTH_TOKEN=<token...>
  
  b) Via configuration file:
  
  $ cachix authtoken <token...>::error::Action failed with error: Error: The process '/home/runner/.nix-profile/bin/cachix' failed with exit code 1
~~~

**Comment by drupol** [comment]
I just re-ran the job, let's see.

**Comment by drupol** [comment]
Would it be possible to add a test? For example a trivial test starting the service and checking if the port is open?

**Comment by drupol** [comment]
I added the tests, let me know if it's good for you.

**Comment by FabianRig** [comment]
> I added the tests, let me know if it's good for you.

Thank you very much! 

That looks good to me!

**Comment by drupol** [comment]
@ofborg build technitium-dns-server

**Comment by FabianRig** [comment]
@ofborg build technitium-dns-server

**Comment by FabianRig** [comment]
> @ofborg build technitium-dns-server

Sorry to bother you again. Anything I can do on my side for this PR?

**Review by drupol — APPROVED** [review]

**Comment by FabianRig** [comment]
Thanks! :-)

**Pull request merged**